node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/OsVersions.java

```diff
@@ -0,0 +1,74 @@
+// Copyright 2018 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+package com.yahoo.vespa.hosted.provision.provisioning;
+
+import com.google.common.collect.ImmutableMap;
+import com.yahoo.component.Version;
+import com.yahoo.config.provision.NodeType;
+import com.yahoo.vespa.curator.Lock;
+import com.yahoo.vespa.hosted.provision.persistence.CuratorDatabaseClient;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.logging.Logger;
+
+/**
+ * Thread-safe class that manages target OS versions for nodes in this repository.
+ *
+ * The target OS version for each node type is set through the /nodes/v2/upgrade REST API.
+ *
+ * @author mpolden
+ */
+public class OsVersions {
+
+    private static final Logger log = Logger.getLogger(OsVersions.class.getName());
+
+    private final CuratorDatabaseClient db;
+
+    /**
+     * Target OS version is read on every request to /nodes/v2/node/[fqdn]. Cache current targets to avoid
+     * unnecessary ZK reads.
+     */
+    private final AtomicReference<ImmutableMap<NodeType, Version>> currentTargets = new AtomicReference<>();
+
+    public OsVersions(CuratorDatabaseClient db) {
+        this.db = db;
+    }
+
+    /** Returns the current target versions for each node type */
+    public Map<NodeType, Version> targets() {
+        return currentTargets.updateAndGet((oldValue) -> oldValue == null ? ImmutableMap.copyOf(db.readOsVersions()) : oldValue);
+    }
+
+    /** Returns the current target version for given node type, if any */
+    public Optional<Version> targetFor(NodeType type) {
+        return Optional.ofNullable(targets().get(type));
+    }
+
+    /** Set the target OS version for nodes of given type */
+    public void setTarget(NodeType nodeType, Version newTarget, boolean force) {
+        if (!nodeType.isDockerHost()) {
+            throw new IllegalArgumentException("Setting target OS version for " + nodeType + " nodes is unsupported");
+        }
+        try (Lock lock = db.lockOsVersions()) {
+            Map<NodeType, Version> osVersions = db.readOsVersions();
+            Optional<Version> oldTarget = Optional.ofNullable(osVersions.get(nodeType));
+
+            if (oldTarget.filter(v -> v.equals(newTarget)).isPresent()) {
+                return; // Old target matches new target, nothing to do
+            }
+
+            if (!force && oldTarget.filter(v -> v.isAfter(newTarget)).isPresent()) {
+                throw new IllegalArgumentException("Cannot set target OS version to " + newTarget +
+                                                   " without setting 'force', as it's lower than the current version: "
+                                                   + oldTarget.get());
+            }
+
+            osVersions.put(nodeType, newTarget);
+            db.writeOsVersions(osVersions);
+            currentTargets.set(ImmutableMap.copyOf(osVersions));
```

Comment:
How will this be updated on other config servers?

Reply:
Good point, will fix.